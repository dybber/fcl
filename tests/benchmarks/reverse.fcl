sig n : int
val n = 4096

sig generateInput : Program <grid> [int]<grid>
val generateInput =
  splitUp 256 (generate (n * n) (fn i => i))
    |> mapPull (fn arr => return <block> (push <block> arr))
    |> concat <block> 256

sig reverse : forall 'a. ['a] -> ['a]
fun reverse arr =
  let n = lengthPull arr
  in generate n (fn i => index arr (n - i - 1))

sig simpleReverse : forall 'a. int -> ['a] -> Program <grid> ['a]<grid>
fun simpleReverse splitSize arr =
  splitUp splitSize arr
   |> mapPull (fn arr2 => return <block> (push <block> (reverse arr2)))
   |> reverse
   |> concat <block> splitSize

sig main : Program <grid> unit
val main =
 do <grid>
  { input <- (bind <grid> generateInput (force <grid>))
  ; arr <- simpleReverse 256 input
  ; forceAndPrint 100 arr
  ; benchmark 100 (bind <grid> (simpleReverse 256 input) (force <grid>))
  }
