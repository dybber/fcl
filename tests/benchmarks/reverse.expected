[16777215,16777214,16777213,16777212,16777211,16777210,16777209,16777208,16777207,16777206,16777205,16777204,16777203,16777202,16777201,16777200,16777199,16777198,16777197,16777196,16777195,16777194,16777193,16777192,16777191,16777190,16777189,16777188,16777187,16777186,16777185,16777184,16777183,16777182,16777181,16777180,16777179,16777178,16777177,16777176,16777175,16777174,16777173,16777172,16777171,16777170,16777169,16777168,16777167,16777166,16777165,16777164,16777163,16777162,16777161,16777160,16777159,16777158,16777157,16777156,16777155,16777154,16777153,16777152,16777151,16777150,16777149,16777148,16777147,16777146,16777145,16777144,16777143,16777142,16777141,16777140,16777139,16777138,16777137,16777136,16777135,16777134,16777133,16777132,16777131,16777130,16777129,16777128,16777127,16777126,16777125,16777124,16777123,16777122,16777121,16777120,16777119,16777118,16777117,16777116]
